### PR TITLE
feat!: Pro Testing Workflow

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR adds in a GitHub action that, on PR and dispatch, will run the most recent version of Semgrep Pro (that is, the one built by the current `develop` branch) with the latest built `semgrep-core` and `semgrep` code.

## Why:
This is important to make sure that we know when we are releasing breaking changes that are incompatible with the latest version of Semgrep Pro! Once we merge this, then we will know that so long as this test passed (the repo is green), and there hasn't been a possibly-breaking `semgrep-proprietary` PR since then (which there shouldn't be, I think), then it is safe to release.

## How:
Added an extra test which will downloaded Semgrep Pro and run it with the just-built version of Semgrep. This means that on PR, the local changes to the PR will be taken into account.

## Test plan:
CI!

Closes PA-2390

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: .github/workflows/semgrep-pro.yml
Comment (IagoAbal) [inline .github/workflows/semgrep-pro.yml]: `--interfile` now

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: This job needs a new name

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: This step shouldn't be necessary, the auth info is only needed for push/pull. You're doing a load from a local file, so you're fine.

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml:81]: Does this command handle installing the deep-semgrep binary? if so, is versioning around that required?

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: Does this semgrep command need access to the `SEMGREP_APP_TOKEN` env var? If so, you'll need to something like `docker run --rm -it -e SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN "returntocorp/semgrep:pr-...` to ensure that env var is propagated into the container.

File: .github/workflows/tests.yml
Comment (spencerdrak) [inline .github/workflows/tests.yml]: This is your call, but if it's not too expensive, we'd strongly recommend running all checks at PR time in addition to merge time. Might be worth removing this.

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: Nit: sub `${{ inputs.repository-name }}` in place of `returntocorp/semgrep`

Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: It looks like `{{ github.event.pull_request.number }}` will be null in cases where it's a workflow-call, workflow-dispatch, or push event. Consider setting up a step at the top with the following logic:
- if it's a pull request, use the PR number
- if it's not a PR or that field is blank, use the `develop` docker image tag

Comment (spencerdrak) [reply .github/workflows/test-semgrep-pro.yaml]: Also this should probably be `${{ ... }}`

**Review by spencerdrak — CHANGES_REQUESTED** [review]

File: .github/workflows/test-semgrep-pro.yaml
Comment (brandonspark) [reply .github/workflows/test-semgrep-pro.yaml]: good call!

File: .github/workflows/test-semgrep-pro.yaml
Comment (brandonspark) [reply .github/workflows/test-semgrep-pro.yaml:81]: nope, my intention is that we try it with the latest `pro` binary, every time

File: .github/workflows/tests.yml
Comment (brandonspark) [reply .github/workflows/tests.yml]: i literally don't even know what this does, so let's do it

File: .github/workflows/test-semgrep-pro.yaml
Comment (brandonspark) [reply .github/workflows/test-semgrep-pro.yaml]: I didn't really know how to do this, but I approximated

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: This isn't going to do what you need it to, this will simply run the docker image (which'll then do whatever the default thing is, I think just a regular semgrep scan) and then will try to run semgrep outside the context of the running docker image. You might want to mount a script into the container, and run that script, or just do something of the form: `docker run ... <args> ... returntocorp/semgrep:pr-123 "semgrep install-semgrep-pro; semgrep --config p/deepsemgrep . --interfile"` 

Not sure that syntax works exactly, but happy to discuss if you want more details.

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml:51]: duplication, is this still needed here?

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]: There's probably better ways to do this, I'll include a short example

Comment (spencerdrak) [reply .github/workflows/test-semgrep-pro.yaml]: https://github.com/returntocorp/semgrep/blob/46bb38e789dce4ac7f4f45cfadf9acbd261b80e1/.github/workflows/release.yml#L35

Try to do something like this, but have the output be the desired tag. Either something of the form `pr-123` or `develop` if you're not on a PR.

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [inline .github/workflows/test-semgrep-pro.yaml]:
```suggestion
          docker run --rm -it -e SEMGREP_APP_TOKEN=${{ secrets.SEMGREP_APP_TOKEN }} --entrypoint bash "${{ inputs.repository_name }}:${{ steps.docker-tag.outputs.docker-tag }}"
```

File: .github/workflows/test-semgrep-pro.yaml
Comment (spencerdrak) [reply .github/workflows/test-semgrep-pro.yaml]: You can set an output like `docker-tag` and use it as suggested below.

**Review by spencerdrak — CHANGES_REQUESTED** [review]

**Comment by brandonspark** [comment]
@spencerdrak ok, it's very close to working, but unfortunately it seems it's having trouble finding the script. could that be because the `checkout` is checking out a particular branch (perhaps `develop`) that doesn't have the script?

File: scripts/test-pro.sh
Comment (spencerdrak) [inline scripts/test-pro.sh:1]: Consider adding `bash -e` here so that failures will short-circuit.

**Review by spencerdrak — APPROVED** [review]
one nit, but overall, this LGTM.

**Comment by brandonspark** [comment]
@spencerdrak seems I cannot merge this until it passes all checks -- do you know how to fix the shellcheck errors?

**Comment by spencerdrak** [comment]
@brandonspark [take a read over this](https://github.com/rhysd/actionlint/blob/main/docs/checks.md) - I believe it's complaining about the unquoted `GITHUB_OUTPUT` bash vars.

**Comment by brandonspark** [comment]
@spencerdrak as they say, ten four. that helped a lot, thanks!

**Pull request merged**
